Support array of bindless textures

icaven opened the pull request:
Currently the library supports single bindless texture handles, but not arrays of bindless texture handles - this PR adds that support.

An example program is added showing how they are used.

icaven committed:
Add support for setting an array of texture handles

icaven committed:
Add an example of using an array of texture handles

icaven committed:
Limit the number of texture handles

Also, remove the setter return value

icaven committed:
Add supports_bindless property to Context

icaven committed:
Update the bindless texture example for the revised directory structure

Commit:
Update README.md TOC automatically

icaven committed:
Merge branch 'moderngl:main' into support-array-of-bindless-textures

icaven committed:
Merge branch 'main'

icaven committed:
Merge branch 'master' into support-array-of-bindless-textures

icaven committed:
Ensure handle list matches array length

icaven committed:
Validate uniform location

icaven committed:
Add runtime check for bindless texture array support

icaven committed:
Fix inconsistent handle length checking

icaven committed:
Encapsulate bindless texture handle management

- Ensure that resident handles are made non-resident before deletion
- Add additional validation for handles

icaven committed:
Add a getter for uniform handles

icaven committed:
Add tests for bindless texture handle arrays

klattimer commented:
I've been testing this branch, and I hit the following error. 

TypeError: argument 3 must be int, not list

klattimer commented:
Nevermind, I had a stale old version getting loaded.

szabolcsdombi commented on src/moderngl.cpp:
```diff
@@ -77,6 +77,49 @@ struct MGLDataType {
+    bool obtained;
+    bool resident;
+
+    BindlessHandleState() : handle(0), obtained(false), resident(false) {}
```
this code will never run see similar parts of the code for initializing with zeros.

szabolcsdombi commented on src/moderngl.cpp:
```diff
@@ -8186,19 +8228,154 @@ static PyObject * MGLContext_write_uniform(MGLContext * self, PyObject * args) {
 
-static PyObject * MGLContext_set_uniform_handle(MGLContext * self, PyObject * args) {
-    int program_obj;
+static PyObject * MGLContext_set_uniform_handle(const MGLContext * self, PyObject *args) {
```
no const here

icaven replied:
removed

icaven committed:
Remove const

icaven committed:
Remove BindlessHandleState constructor

This code would never have been run, and the call to reset() initializes the struct

icaven committed:
Restore README.md to match main

icaven committed:
Reject scalar Uniform.handle on array uniform

The setter previously accepted a scalar for an array uniform and
silently wrote only slot 0 via ProgramUniformHandleui64ARB. Almost
always a "forgot to wrap in a list" bug. Raise ValueError instead,
matching the existing list-length check.

icaven committed:
Use portable shader version in bindless tests

Switch all #version 440 directives to #version 330 in
tests/test_bindless_handles.py. Fragment shaders already had
#extension GL_ARB_bindless_texture : require; vertex shaders don't
use bindless features. NVIDIA's permissive GLSL compiler accepted
440 in a 3.3 context, but spec-conformant compilers (AMD, Intel,
Mesa) reject it. Matches the pattern used by the existing examples.

icaven committed:
Document BindlessHandleState init contract

Explain why there's no constructor (PyObject_New bypasses C++ ctors)
and that every alloc site must call bindless_state.reset() before
the fields are read.

icaven committed:
Validate array_length in _get_uniform_handle

glGetUniformui64vARB writes all elements of a uniform array into the
caller's buffer with no count parameter, so a caller-supplied
array_length smaller than the program's actual array size overflows
our PyMem_Malloc'd buffer.

Add lookup_uniform_array_size to query the actual size from the
program's active uniforms, and reject any mismatch in
MGLContext_get_uniform_handle. Closes the heap-overflow window even
when callers bypass the Python wrapper (e.g. by mutating
Uniform.array_length or calling _get_uniform_handle directly on the
mglo).

icaven committed:
Match codebase style in bindless code

- Switch PyObject*Foo to PyObject * Foo, static_cast<> to C-style
  casts, and auto* to explicit GLuint64 *.
- Trim the double blank line in BindlessHandleState left by the
  removed constructor.
- Rename test_handle_negative_location to
  test_nonexistent_uniform_not_in_program and drop the irrelevant
  bindless skip; the body doesn't use bindless.
- Convert the gated `if array_length == 1` in
  test_handle_uniform_not_array into an assertion.

icaven committed:
Test bindless residency state machine and lifecycle

- Add a comment in test_single_handle_backward_compatibility
  explaining why release-after-assign is safe (handle is made
  non-resident before deletion) and warning not to sample from
  the dangling uniform value afterward.
- Add test_bindless_state_machine_no_gl_errors that walks through
  the BindlessHandleState transitions (memoized get_handle,
  redundant set_residency, release-while-resident, release without
  ever obtaining a handle, release without ever being resident).
  After each step it asserts both ctx.error == "GL_NO_ERROR" and
  glIsTextureHandleResidentARB(handle) matches the expected state.
  The residency-query function pointer is loaded via
  ctx.mglo._context.load_opengl_function + ctypes since PyOpenGL's
  static binding doesn't resolve ARB extension functions against
  EGL contexts.

icaven committed:
Require extension for supports_bindless

GL_ARB_bindless_texture is not in core OpenGL even at 4.6 -- the
extension must be exposed by the implementation. The previous
version_code >= 440 shortcut was a false positive on Mesa's software
rasterizers, which report a high GL version but don't implement
bindless. Drop the shortcut and rely solely on the extension presence.

Replace the test that asserted the buggy behavior with a regression
test pinning that no GL version implies bindless support without the
extension.

icaven committed:
Run wrapper validation tests on non-bindless contexts

Several tests in test_bindless_handles.py exercise Python-side
validation in Uniform.handle (list length, scalar-to-array) and
C-side pre-GL guards in _set_uniform_handle / _get_uniform_handle
(type, overflow, array_length mismatch). Those checks all fire
before any bindless GL call, so they only need a uniform with a
known array_length -- not a working bindless context.

Add a make_sampler_array_program(count) factory fixture that
produces a minimal #version 330 program with a regular (non-bindless)
sampler2D Textures[count] uniform, drop the supports_bindless skip
from the eight applicable tests, and replace tex.get_handle() calls
with dummy integers since the validation path never reaches GL.

This lets the wrapper-logic tests run on CI runners that don't
support GL_ARB_bindless_texture (e.g. Mesa), while end-to-end tests
that need real bindless residency keep their skip.
